File: three_viz/dog/environment.js
/** Continuous dog game environment (matches doggame/environment.py). */

const ANGLES_DEG = [
  0, 22.5, 45, 67.5, 90, 112.5, 135, 157.5,
  180, 202.5, 225, 247.5, 270, 292.5, 315, 337.5,
];

export const NUM_ACTIONS = 17;
export const STEP_SIZE = 0.1;
export const DEFAULT_HORIZON = 30;
export const DEFAULT_HOUSE1 = [0.25, 0.25];
export const DEFAULT_HOUSE2 = [0.75, 0.75];
/** Capsule radius in state coords — matches 3D pill on the 1×1 arena. */
export const PLAYER_RADIUS = 0.022;

export const ACTION_NAMES = [
  "Stay", "E", "ENE", "NE", "NNE", "N", "NNW", "NW", "WNW",
  "W", "WSW", "SW", "SSW", "S", "SSE", "SE", "ESE",
];

function buildActionDirs() {
  const dirs = { 0: [0, 0] };
  for (let i = 0; i < ANGLES_DEG.length; i++) {
    const rad = (ANGLES_DEG[i] * Math.PI) / 180;
    dirs[i + 1] = [Math.cos(rad), Math.sin(rad)];
  }
  return dirs;
}

const ACTION_DIRS = buildActionDirs();

export function dogPosition(state) {
  return [(state[0] + state[2]) / 2, (state[1] + state[3]) / 2];
}

function distance(p1, p2) {
  const dx = p1[0] - p2[0];
  const dy = p1[1] - p2[1];
  return Math.sqrt(dx * dx + dy * dy);
}

export function clampPosition(x, y) {
  const lo = PLAYER_RADIUS;
  const hi = 1 - PLAYER_RADIUS;
  return [
    Math.max(lo, Math.min(hi, x)),
    Math.max(lo, Math.min(hi, y)),
  ];
}

export function createDogEnv(options = {}) {
  const stepSize = options.stepSize ?? STEP_SIZE;
  const house1 = options.house1 ?? DEFAULT_HOUSE1;
  const house2 = options.house2 ?? DEFAULT_HOUSE2;

  function move(x, y, action) {
    const [dx, dy] = ACTION_DIRS[action];
    return clampPosition(x + dx * stepSize, y + dy * stepSize);
  }

  return {
    stepSize,
    house1,
    house2,
    numActions: NUM_ACTIONS,

    transition(state, a1, a2) {
      const [x1, y1] = move(state[0], state[1], a1);
      const [x2, y2] = move(state[2], state[3], a2);
      return [x1, y1, x2, y2];
    },

    /** Returns separate rewards (r1, r2) after the step. */
    reward(state, a1, a2) {
      const next = this.transition(state, a1, a2);
      const dog = dogPosition(next);
      return {
        r1: -distance(dog, house1),
        r2: -distance(dog, house2),
      };
    },

    sampleState() {
      const lo = PLAYER_RADIUS;
      const span = 1 - 2 * PLAYER_RADIUS;
      return [
        lo + Math.random() * span,
        lo + Math.random() * span,
        lo + Math.random() * span,
        lo + Math.random() * span,
      ];
    },
  };
}

File: doggame/environment.py
import random

import numpy as np

try:
    from .config import ACTION_DIRS, DEFAULT_HOUSE1, DEFAULT_HOUSE2
except ImportError:
    from config import ACTION_DIRS, DEFAULT_HOUSE1, DEFAULT_HOUSE2


def dog_position(s):
    """Dog is at the midpoint of the two players."""
    return ((s[0] + s[2]) / 2, (s[1] + s[3]) / 2)


def distance(p1, p2):
    """Euclidean distance between two points."""
    return np.sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)


PLAYER_RADIUS = 0.022


class DogGame:
    """
    Dog Game Environment.

    State: (p1_x, p1_y, p2_x, p2_y) - all in [0, 1]
    Dog: midpoint of players
    Reward: -distance(dog, my_house)
    """

    def __init__(self, step_size=0.1, house1=None, house2=None):
        self.step_size = step_size
        self.house1 = house1 if house1 else DEFAULT_HOUSE1
        self.house2 = house2 if house2 else DEFAULT_HOUSE2

    def move(self, x, y, action):
        """Apply action and clamp so the player disc stays inside [0, 1]²."""
        dx, dy = ACTION_DIRS[action]
        lo, hi = PLAYER_RADIUS, 1 - PLAYER_RADIUS
        x_new = np.clip(x + dx * self.step_size, lo, hi)
        y_new = np.clip(y + dy * self.step_size, lo, hi)
        return x_new, y_new

    def transition(self, s, a1, a2):
        """Return next state after both players move."""
        x1_new, y1_new = self.move(s[0], s[1], a1)
        x2_new, y2_new = self.move(s[2], s[3], a2)
        return (x1_new, y1_new, x2_new, y2_new)

    def reward(self, s, a1, a2):
        """Compute rewards for both players."""
        s_next = self.transition(s, a1, a2)
        dog = dog_position(s_next)
        r1 = -distance(dog, self.house1)
        r2 = -distance(dog, self.house2)
        return r1, r2

    def sample_state(self):
        """Sample a uniform random state with room for player radius at edges."""
        lo = PLAYER_RADIUS
        span = 1 - 2 * PLAYER_RADIUS
        return (
            lo + random.random() * span,
            lo + random.random() * span,
            lo + random.random() * span,
            lo + random.random() * span,
        )
